packages/effect/src/TSubscriptionRef.ts

```diff
@@ -0,0 +1,171 @@
+/**
+ * @since 2.0.0
+ */
+import * as internal from "./internal/stm/tSubscriptionRef.js"
+import type * as Option from "./Option.js"
+import type { Pipeable } from "./Pipeable.js"
+import type * as STM from "./STM.js"
+import type * as Stream from "./Stream.js"
+import type * as TPubSub from "./TPubSub.js"
+import type * as TRef from "./TRef.js"
+import type * as Types from "./Types.js"
+
+/**
+ * @since 2.0.0
+ * @category symbols
+ */
+export const TSubscriptionRefTypeId: unique symbol = internal.TSubscriptionRefTypeId
+
+/**
+ * @since 2.0.0
+ * @category symbols
+ */
+export type TSubscriptionRefTypeId = typeof TSubscriptionRefTypeId
+
+/**
+ * A `TSubscriptionRef<A>` is a `TRef` that can be subscribed to in order to
+ * receive the current value as well as a `Stream` of all changes to the value.
+ *
+ * @since 2.0.0
+ * @category models
+ */
+export interface TSubscriptionRef<in out A> extends TSubscriptionRef.Variance<A>, Pipeable {
+  /** @internal */
+  readonly ref: TRef.TRef<A>
+  /** @internal */
+  readonly pubsub: TPubSub.TPubSub<A>
+  /**
+   * A stream containing the current value of the `Ref` as well as all changes
+   * to that value.
+   */
+  readonly changes: Stream.Stream<A>
```

Comment:
I think this should be a `STM<TQueue<A>>`, and maybe also add a `changesScoped: Effect<TQueue<A>, never, Scope>` and `changesStream`.

The scoped and stream variants could just be module functions.

Reply:
Sorry for taking so long to get to this. I'm having a bit of trouble figuring out how to do this with a TQueue. The issue is that in order to provide a TQueue that has the current value of the ref and any changes to it I'd need some way to merge queues. Like this for example does not work:

```ts
 get changes(): STM.STM<TQueue.TQueue<A>>{
    return STM.flatMap(TQueue.unbounded<A>(), queue => pipe(
      STM.flatMap(TRef.get(this.ref), a => TQueue.offer(queue, a)),
      STM.flatMap(() => TPubSub.subscribe(this.pubsub)),
      STM.flatMap(dequeue => pipe(
          TQueue.poll(dequeue),
          STM.tap(a => Option.isNone(a) ? TQueue.shutdown(dequeue) : TQueue.offer(queue, a.value)),
      )),
      STM.as(queue)
    )
```

I don't see how I can provide a TQueue with the current value _and_ any changes from the queue given by `TPubSub.subscribe`. Any thoughts or pointers?

Reply:
changes doesn't contain the current value usually

Reply:
@mikearnaldi it does with SubscriptionRef which is why I thought we probably wanted the same behavior here, from SubscriptionRef.ts:

```ts
 /**
   * A stream containing the current value of the `Ref` as well as all changes
   * to that value.
   */
  readonly changes: Stream.Stream<A>
```

Reply:
then probably just implementing the TDequeue interface with a custom impl

Reply:
I think thanks to STM transactional behaviour we can actually merge queues: https://github.com/Effect-TS/effect/pull/2725/commits/91e6a5552fe1e917db730c6ed1285e949f2ad4ea this would need a bit of testing though